src/java/org/apache/cassandra/service/StorageProxy.java

```diff
@@ -2262,26 +2286,38 @@ public static boolean shouldHint(Replica replica, boolean tryEnablePersistentWin
         long endpointDowntime = Gossiper.instance.getEndpointDowntime(endpoint);
         boolean hintWindowExpired = endpointDowntime > maxHintWindow;
 
-        if (tryEnablePersistentWindow && !hintWindowExpired && DatabaseDescriptor.hintWindowPersistentEnabled())
+        UUID hostIdForEndpoint = StorageService.instance.getHostIdForEndpoint(endpoint);
+        if (hostIdForEndpoint == null)
         {
-            UUID hostIdForEndpoint = StorageService.instance.getHostIdForEndpoint(endpoint);
-            if (hostIdForEndpoint != null)
-            {
-                long earliestHint = HintsService.instance.getEarliestHintForHost(hostIdForEndpoint);
-                hintWindowExpired = Clock.Global.currentTimeMillis() - maxHintWindow > earliestHint;
-                if (hintWindowExpired)
-                    Tracing.trace("Not hinting {} for which there is the earliest hint stored at {}", replica, earliestHint);
-            }
+            Tracing.trace("Discarding hint for endpoint not part of ring: {}", endpoint);
+            return false;
         }
-        else if (hintWindowExpired)
+
+        // if persisting hints window, hintWindowExpired might be updated according to the timestamp of the earliest hint
+        if (tryEnablePersistentWindow && !hintWindowExpired && DatabaseDescriptor.hintWindowPersistentEnabled())
         {
-            Tracing.trace("Not hinting {} which has been down {} ms", replica, endpointDowntime);
+            long earliestHint = HintsService.instance.getEarliestHintForHost(hostIdForEndpoint);
+            hintWindowExpired = Clock.Global.currentTimeMillis() - maxHintWindow > earliestHint;
+            if (hintWindowExpired)
+                Tracing.trace("Not hinting {} for which there is the earliest hint stored at {}", replica, earliestHint);
         }
 
         if (hintWindowExpired)
-            HintsService.instance.metrics.incrPastWindow(replica.endpoint());
+        {
+            HintsService.instance.metrics.incrPastWindow(endpoint);
+            Tracing.trace("Not hinting {} which has been down {} ms", endpoint, endpointDowntime);
+            return false;
+        }
+
+        long maxHintsSize = DatabaseDescriptor.getMaxHintsSizePerHost();
+        boolean hasHintsReachedMaxSize = maxHintsSize > 0 && HintsService.instance.getTotalHintsSize(hostIdForEndpoint) > maxHintsSize;
+        if (hasHintsReachedMaxSize)
+        {
+            Tracing.trace("Not hinting {} which has reached to the max hints size {} bytes on disk", endpoint, DatabaseDescriptor.getMaxHintsSizePerHost());
```

Comment:
Maybe also log the total hint size value here: `HintsService.instance.getTotalHintsSize(hostIdForEndpoint)`. It might come in handy for debugging.